### PR TITLE
SNOW-2037930:Fix copy into with case insensitive option having the variant size limitation

### PR DESCRIPTION
**Pull request by sfc-gh-yuwang**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-NNNNNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)
   - [x] If adding any arguments to public Snowpark APIs or creating new public Snowpark APIs, I acknowledge that I have ensured my changes include AST support. Follow the link for more information: [AST Support Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#ast-abstract-syntax-tree-support-in-snowpark)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/data_source/utils.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ def _upload_and_copy_into_table(
     copy_into_table_query = f"""
     COPY INTO {snowflake_table_name} FROM @{snowflake_stage_name}/{file_name}
     FILE_FORMAT = (TYPE = PARQUET USE_VECTORIZED_SCANNER=TRUE)
-    MATCH_BY_COLUMN_NAME=CASE_INSENSITIVE
+    MATCH_BY_COLUMN_NAME=CASE_SENSITIVE
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/utils.py:179]: can we have a test for the case sensitivity?

File: src/snowflake/snowpark/_internal/data_source/drivers/base_driver.py [hunk 2]
```diff
@@ -162,6 +168,15 @@ def data_source_data_to_pandas_df(
...
             lambda x: x.hex() if isinstance(x, (bytearray, bytes)) else x
         )
+
+        # when a column have all None data and corresponding type in snowflake is BinaryType or BooleanType,
+        # generated parquet file will have a physical type of INT32 that is not compatible with copy into,
+        # this piece of code convert those type to string type to avoid the issue, and this is a safe change
+        # since convert these two type to string will not lose any precision or information
+        for field in schema.fields:
+            if isinstance(field.datatype, (BinaryType, BooleanType)):
+                df[field.name] = df[field.name].astype("string")
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/data_source/drivers/base_driver.py:178]: can you test some corner cases?

for example, if I have a pandas dataframe with boolean data

```python
import pandas as pd
df = pd.DataFrame({'flag': [False]})
df['flag'] = df['flag'].astype("string")

# now df flag becomes all string
if df['flag'][1]:  # it's now checking `False` string
    print('condition is true')
```
